### ISSUE TITLE
azcopy can't resume cloning job if controller pod is killed during job run

### ISSUE THREAD

**Issue by RomanBednar** [kubernetes-sigs/azurefile-csi-driver#1950]
**What happened**:

While testing azure file cloning in OpenShift we noticed that if a controller pod running azcopy job is killed any clone PVC still being copied gets stuck in Pending phase.

This seems to be the effect of azcopy relying on job plan files (`AZCOPY_JOB_PLAN_LOCATION` or `~/.azcopy` by default) to track jobs which can be lost if stored on ephemeral volume.

Is this a know limitation or is there a recommended solution?

**What you expected to happen**:

Clone job surviving a lost controller pod.

**How to reproduce it**:

1) Create Azure File PVC/PV 
2) Create a new clone PVC referencing the origin PVC as a source
3) Kill Azure File leader controller pod
4) Clone PVC is stuck in Pending state

**Anything else we need to know?**:

Checking the helm charts in this repo the destination for those job plan files seems to be ephemeral with `emptyDir` volume so the issue would occur with this deployment as well: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/eefed914745603ad28eac52f07c47b972ad3dadd/charts/v1.30.2/azurefile-csi-driver/templates/csi-azurefile-controller.yaml#L229

**Environment**:

- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by andyzhangx** [comment]
yes, that's a limitation since the job status can only be stored locally in `~/.azcopy`, as long as the controller pod is restarted, copy jobs could not be continued.

**Comment by jsafrane** [comment]
How can users detect such state? How can they recover / resume?

Who deletes incompletely cloned volumes in Azure? There is no PV created for them. How is an user supposed to find them in the first place?

I am afraid the cloning is not very useful if it can leak volumes that need manual cleanup.

**Comment by RomanBednar** [comment]
@andyzhangx We're considering moving snapshots and cloning to GA in the next release - is this limitation still present?

**Comment by andyzhangx** [comment]
> [@andyzhangx](https://github.com/andyzhangx) We're considering moving snapshots and cloning to GA in the next release - is this limitation still present?

@RomanBednar 
we have increased the memory and cpu limit, have done following two large scale tests which shows test could pass:
 - 125K file num in one file share( each 3MB file) (takes ~21min)
 - 10TB Large Sized File Share with large file counts (each 1GB file)  (takes ~15hours)

If your data size is within the above limit, that should be fine.

**Comment by RomanBednar** [comment]
@andyzhangx Seems like the limit testing was related to different issue? What I had in mind was what we originally discussed here some time ago which is that there was no recovery when the controller pod holding the azcopy logs gets deleted during some longer running operation like cloning a volume - even if the volume is small it can get interrupted.

**Comment by andyzhangx** [comment]
> [@andyzhangx](https://github.com/andyzhangx) Seems like the limit testing was related to different issue? What I had in mind was what we originally discussed here some time ago which is that there was no recovery when the controller pod holding the azcopy logs gets deleted during some longer running operation like cloning a volume - even if the volume is small it can get interrupted.

@RomanBednar if controller pod is restarted, the azcopy would perform the volume cloning from scratch. and if the job takes very long time, and controller pod is restarted during the volume clone, that would make the job taker longer time, that's the issue, so there is limit in azcopy solution, and as long as the workload of your scenario does not exceed the above testing scenario, that should be ok.

**Comment by RomanBednar** [comment]
@andyzhangx Thanks for the details, does that mean that azcopy does not rely on the job/plan files to resume the operation or tracking the limit you mentioned? Because if using [emptyDir](https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/3c7d8bb1d792ef72a096ac725061afaa63d57cae/deploy/v1.33.3/csi-azurefile-controller.yaml#L195) all the files are gone along with the pod.

I tried to find relevant azcopy PRs too see how this changed since I last tested this, but to no avail - so any pointers are welcome too. I also believe our docs team would like to document the limits you mentioned, is this published anywhere so we can use it as a reference?

**Comment by andyzhangx** [comment]
> [@andyzhangx](https://github.com/andyzhangx) Thanks for the details, does that mean that azcopy does not rely on the job/plan files to resume the operation or tracking the limit you mentioned? Because if using [emptyDir](https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/3c7d8bb1d792ef72a096ac725061afaa63d57cae/deploy/v1.33.3/csi-azurefile-controller.yaml#L195) all the files are gone along with the pod.
> 
> I tried to find relevant azcopy PRs too see how this changed since I last tested this, but to no avail - so any pointers are welcome too. I also believe our docs team would like to document the limits you mentioned, is this published anywhere so we can use it as a reference?

as long as the controller pod is restarted, the azcopy will perform the operation from the beginning. We have only tested the above scenarios since that's P99 users' scenarios, we don't have published that in anywhere else, I think you could perform the tests on your side. @RomanBednar

**Comment by RomanBednar** [comment]
@andyzhangx Alright, thanks for the details! I've tested this scenario with clones and snapshot restore and the azcopy jobs were able to recover. We still need to do planning but I think nothing is blocking us now from moving snaps and clones to GA in 4.21.
